File: design_pattern/src/Ex1/BookShelf.java
package Ex1;
import java.util.*;

public class BookShelf implements Aggregate {
//	private Book[] books;
	private ArrayList<Book> books = new ArrayList<Book>();
	private int last = 0;
	public BookShelf(int maxsize) {
//		this.books = new Book[maxsize];
		this.books = new ArrayList<Book>();
	}
	public Book getBookAt(int index) {
//		return books[index];
		return (Book)books.get(index);
	}
	public void appendBook(Book book) {
//		this.books[last] = book;
		this.books.add(book);
		last++;
	}
	public int getLength() {
		return last;
	}
	public Iterator iterator() {
		return new BookShelfIterator(this);
	}
}

File: python/iterator/book_shelf.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

from aggregate import Aggregate
from book_shelf_iterator import BookShelfIterator

class BookShelf(Aggregate):
    def __init__(self, maxsize):
        self.__last = 0
        self.__books = [0 for i in range(maxsize)]

    def get_book_at(self, index):
        return self.__books[index]

    def append_book(self, book):
        self.__books[self.__last] = book
        self.__last += 1

    def get_length(self):
        return self.__last

    def iterator(self):
        return BookShelfIterator(self)
